### PR TITLE
Praateek

### PR DESCRIPTION
**Pull request by praateekmahajan**
Added whole lot of support for funnel, email subscriber scheduler and mandrill as well.

### CODE REVIEW

File: octo-core/lib/octocore/mailer/subscriber_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    class SubscriberMailer
+      @queue = :subscriber_notifier
+
+      def perform (from=nil)
+        if from.nil?
+          subscribers = Octo::Subscriber.where(created_at: 24.hours.ago..Time.now.floor)
+        else
+          subscribers = Octo::Subscriber.where(created_at: from..Time.now.floor)
+        end
+        puts subscribers.length
```
Comment (pranny) [inline octo-core/lib/octocore/mailer/subscriber_mailer.rb]: Why puts here?

File: octo-core/lib/octocore/mailer/subscriber_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        #   MAIL CODE
```
Comment (pranny) [inline octo-core/lib/octocore/mailer/subscriber_mailer.rb:19]: Keep the code as it is.

File: octo-core/bin/fakestream [hunk 3]
```diff
@@ -40,20 +37,25 @@ module Octo
 
       # simulate an app login
       app_init user
-
+      # sleep 0.5
```
Comment (pranny) [inline octo-core/bin/fakestream]: Remove if not necessary. Applies to other instances as well.

File: octo-core/lib/octocore.rb [hunk 4]
```diff
@@ -3,17 +3,22 @@
 require 'logger'
 
 require 'octocore/version'
-require 'octocore/utils'
 require 'octocore/config'
 require 'octocore/models'
 require 'octocore/counter'
+# require 'octocore/email'
```
Comment (pranny) [inline octo-core/lib/octocore.rb]: Remove if not necessary

File: octo-core/lib/octocore.rb [hunk 5]
```diff
@@ -56,14 +61,15 @@ def self.connect_with_config_dir(config_dir)
...
+
     # Now, good to merge the two
     self.connect config
   end
 
-  # Loads the true config. The true config is the hierarchial config
-  # @param [Hash] config The base config. Loaded config will be deep merged
-  #   with this
-  # @param [String] file The file from which config should be loaded
+# Loads the true config. The true config is the hierarchial config
```
Comment (pranny) [inline octo-core/lib/octocore.rb]: Why change the indentation here?

File: octo-core/lib/octocore/callbacks.rb [hunk 6]
```diff
@@ -15,6 +15,9 @@ module OctoHooks
 
       # Define the after_app_init hook
       after_app_init do |args|
+        # Not sure if we need to add init to
```
Comment (pranny) [inline octo-core/lib/octocore/callbacks.rb]: Remove if not needed.

File: octo-core/lib/octocore/callbacks.rb [hunk 7]
```diff
@@ -64,15 +69,28 @@ def update_counters(opts)
...
+      # product_page_view to redis.
+      # It self expires in n seconds from last hit
+      def add_session(opts)
+        if opts.has_key?(:type)
+          # create a shadow key for every list
+          # this helps us in catching the event
+          # when a key expires, and then we
+          # delete the main key with the session details
+          # after playing around with the value of the key
+          Cequel::Record.redis.setex("shadow:" + opts[:enterprise].id.to_s + "_" + opts[:user].id.to_s,Octo.get_config(:session_length),"")
```
Comment (pranny) [inline octo-core/lib/octocore/callbacks.rb]: Create a method that creates this shadow key.

File: octo-core/lib/octocore/helpers/api_consumer_helper.rb [hunk 8]
```diff
@@ -110,8 +120,27 @@ def call_hooks(event, *args)
         hook = [:after, event.gsub('.', '_')].join('_').to_sym
         Octo::Callbacks.run_hook(hook, *args)
       end
+      def checkFunnelTracker(enterprise,page1,page2)
```
Comment (pranny) [inline octo-core/lib/octocore/helpers/api_consumer_helper.rb]: A space above this method and some comments would be nice.

File: octo-core/lib/octocore/helpers/api_consumer_helper.rb [hunk 9]
```diff
@@ -110,8 +120,27 @@ def call_hooks(event, *args)
...
+        }
+        counters = {
+            weight:1,
+        }
+        Octo::FunnelTracker.findOrCreateOrAdjust(args_to,counters)
+        Octo::FunnelTracker.findOrCreateOrAdjust(args_from,counters)
+      end
 
-      def checkUserProfileDetails(enterprise, user, msg)
+        def checkUserProfileDetails(enterprise, user, msg)
```
Comment (pranny) [inline octo-core/lib/octocore/helpers/api_consumer_helper.rb]: Why change the indentation?

File: octo-core/lib/octocore/scheduler.rb [hunk 10]
```diff
@@ -54,6 +54,18 @@ def schedule_baseline
         end
       end
 
+      def schedule_subscribermail
```
Comment (pranny) [inline octo-core/lib/octocore/scheduler.rb:58]: Some comments would be really helpful.

File: octo-core/lib/octocore/utils.rb [hunk 11]
```diff
@@ -22,7 +22,7 @@ def deserialize(data)
   end
 end
 
-class ::Time
+class Time
```
Comment (pranny) [inline octo-core/lib/octocore/utils.rb]: Why revert the changes of this file? Please keep all the changes on the left.

File: octo-core/lib/octocore/version.rb [hunk 12]
```diff
@@ -1,4 +1,4 @@
 module Octo
   # The current version of the library
-  VERSION = '0.0.2'
+  VERSION = '0.0.1'
```
Comment (pranny) [inline octo-core/lib/octocore/version.rb]: Keep the version as 0.0.2

File: octo-core/lib/octocore.rb [hunk 13]
```diff
@@ -56,14 +61,15 @@ def self.connect_with_config_dir(config_dir)
     # with the global config hash
     configConfig = config.delete(:config)
     config = config.deep_merge(configConfig)
+
```
Comment (pranny) [inline octo-core/lib/octocore.rb]: Have you added space here?

File: octo-core/lib/octocore/callbacks.rb (hunk 6 above)
Comment (pranny) [inline octo-core/lib/octocore/callbacks.rb]: I think you should keep the session upon this call.

File: octo-core/lib/octocore/callbacks.rb (hunk 7 above)
Comment (pranny) [inline octo-core/lib/octocore/callbacks.rb]: Use single quotes when the statement inside is pure string literal and wont be evaluated.

File: octo-core/lib/octocore/helpers/api_consumer_helper.rb [hunk 14]
```diff
@@ -14,89 +14,99 @@ def valid_events
...
+                call_hooks(eventName, hook_opts)
+              when 'update.profile'
+                checkUserProfileDetails(enterprise, user, msg)
+                updateUserDeviceDetails(user, msg)
+                call_hooks(eventName, hook_opts)
+              when 'update.push_token'
+                checkPushToken(enterprise, user, msg)
+                checkPushKey(enterprise, msg)
+              when 'funnel_update'
+                sessionList = Cequel::Record.redis.lrange(msg[:rediskey],0,-1)
```
Comment (pranny) [inline octo-core/lib/octocore/helpers/api_consumer_helper.rb]: Put this in a method and call it here.

File: octo-core/lib/octocore/mailer/subscriber_mailer.rb [hunk 15]
```diff
@@ -0,0 +1,26 @@
+require 'octocore'
+module Octo
+  module Mailer
+    class SubscriberMailer
```
Comment (pranny) [inline octo-core/lib/octocore/mailer/subscriber_mailer.rb:5]: A comments would be appreciated.

File: octo-core/lib/octocore/models/enterprise/funnel_tracker.rb [hunk 16]
```diff
@@ -0,0 +1,14 @@
+require 'cequel'
+
+module Octo
+  class FunnelTracker
```
Comment (pranny) [inline octo-core/lib/octocore/models/enterprise/funnel_tracker.rb:9]: A few comments about the tracker class and the data structure it encapsulates. Also this class should provide helper functions like `incoming_nodes`, `outgoing_nodes` , `load_factor` etc. These would be helpful in computing markov chains later.

**Comment by pranny** [comment]
Apart from the comments above, it looks good to me to gathering data. Now, you would also have to create markov chains based on this data. And compute statistics like top chains per day or real time.

**Comment by pranny** [comment]
Apart from the comments above, it looks good to me to gathering data. Now, you would also have to create markov chains based on this data. And compute statistics like top chains per day or real time.

File: octo-core/lib/octocore/mailer/subscriber_mailer.rb (hunk 2 above)
Comment (praateekmahajan) [inline octo-core/lib/octocore/mailer/subscriber_mailer.rb:19]: Didn't get this one...

File: octo-core/lib/octocore.rb (hunk 4 above)
Comment (praateekmahajan) [inline octo-core/lib/octocore.rb]: We need it. It is the functionality Ravin Bhaiya made. Although it is leading to weird error " **mandril** not found", so probably once integrated with master this error has to be looked at.

File: octo-core/lib/octocore.rb (hunk 5 above)
Comment (praateekmahajan) [inline octo-core/lib/octocore.rb]: Every change in indentation is due to reason unknown, nothing intentional.

File: octo-core/lib/octocore/callbacks.rb (hunk 6 above)
Comment (praateekmahajan) [inline octo-core/lib/octocore/callbacks.rb]: You mean, init to init be one session? But we already got that covered with the expiry limit..

File: octo-core/lib/octocore/callbacks.rb (hunk 7 above)
Comment (praateekmahajan) [inline octo-core/lib/octocore/callbacks.rb]: Done

File: octo-core/lib/octocore/mailer/subscriber_mailer.rb (hunk 1 above)
Comment (praateekmahajan) [inline octo-core/lib/octocore/mailer/subscriber_mailer.rb]: It was there, for debugging purpose, forgot to remove. Done :+1:

File: octo-core/lib/octocore.rb [hunk 17]
```diff
@@ -60,10 +65,10 @@ def self.connect_with_config_dir(config_dir)
     self.connect config
   end
 
-  # Loads the true config. The true config is the hierarchial config
-  # @param [Hash] config The base config. Loaded config will be deep merged
-  #   with this
-  # @param [String] file The file from which config should be loaded
+# Loads the true config. The true config is the hierarchial config
+# @param [Hash] config The base config. Loaded config will be deep merged
```
Comment (pranny) [inline octo-core/lib/octocore.rb]: I think its your editor. Make sure your editor replaces tab by 2 spaces. Comments should have the same indentation level as the method http://www.rubydoc.info/gems/yard/file/docs/Tags.md#List_of_Available_Tags

File: octo-core/lib/octocore.rb (hunk 4 above)
Comment (pranny) [inline octo-core/lib/octocore.rb]: Can you try and fix the error? Maybe it's a gem dependency that is not mentioned in `octocore.gemspec` or maybe the spelling is wrong. The correct spelling is **Mandrill** iirc

File: octo-core/lib/octocore/mailer/subscriber_mailer.rb [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+      # Counts the number of subscriber in the last 24 hours
+      #  and then sends a mail with subscriber count to the
+      #  email mentioned
+
+      def perform (from=nil)
+        if from.nil?
+          subscribers = Octo::Subscriber.where(created_at: 24.hours.ago..Time.now.floor)
+        else
+          subscribers = Octo::Subscriber.where(created_at: from..Time.now.floor)
+        end
```
Comment (pranny) [inline octo-core/lib/octocore/mailer/subscriber_mailer.rb:18]: Please uncomment this. The functionality should be present in the code. You could add a check on top level `Octo::Email.send` to send email only if the required keys are present. Remaining part of the code should be abstracted out of this thing. This goes in line with the require 'email' comment.

File: octo-core/lib/octocore/models/enterprise/funnel_tracker.rb [hunk 19]
```diff
@@ -0,0 +1,19 @@
+require 'cequel'
+
+# A model for tracking the user web flow
```
Comment (pranny) [inline octo-core/lib/octocore/models/enterprise/funnel_tracker.rb]: This comment should come before `class Funnel..` class definition. Else it would be generated against Octo module.

File: octo-core/lib/octocore/helpers/api_consumer_helper.rb [hunk 20]
```diff
@@ -111,6 +114,51 @@ def call_hooks(event, *args)
         Octo::Callbacks.run_hook(hook, *args)
       end
 
+      # Checks for msg[:rediskey] in redis, parses it and
+      #  then calls updateFunnelTracker.
+      # @param [Octo::Enterprise] enterprise The Enterprise object
+      # @param [Hash] msg The message hash, MUST contain, :rediskey
+      # @return [void]
+
```
Comment (pranny) [inline octo-core/lib/octocore/helpers/api_consumer_helper.rb]: No spaces between comments and method defns
